controllers/controllers/clusters/vsphere.go

```diff
@@ -2,19 +2,153 @@ package clusters
+	}
+
+	if err := v.Defaulter.SetDefaultsForDatacenterConfig(ctx, vsphereDatacenter); err != nil {
+		return fmt.Errorf("failed setting default values for vsphere datacenter config: %v", err)
+	}
+
+	return nil
+}
+
+func (v *VSphereClusterReconciler) bundles(ctx context.Context, name, namespace string) (*releasev1alpha1.Bundles, error) {
+	clusterBundle := &releasev1alpha1.Bundles{}
+	err := v.capiResourceFetcher.FetchObjectByName(ctx, name, namespace, clusterBundle)
+	if err != nil {
+		return nil, err
+	}
+	return clusterBundle, nil
+}
+
+func (v *VSphereClusterReconciler) FetchAppliedSpec(ctx context.Context, cs *anywherev1.Cluster) (*c.Spec, error) {
+	return c.BuildSpecForCluster(ctx, cs, v.bundles, nil)
+}
+
+func (v *VSphereClusterReconciler) Reconcile(ctx context.Context, cluster *anywherev1.Cluster) (reconciler.Result, error) {
+	dataCenterConfig := &anywherev1.VSphereDatacenterConfig{}
+	dataCenterName := types.NamespacedName{Namespace: cluster.Namespace, Name: cluster.Spec.DatacenterRef.Name}
+	if err := v.Client.Get(ctx, dataCenterName, dataCenterConfig); err != nil {
+		return reconciler.Result{}, err
+	}
+	// Set up envs for executing Govc cmd and default values for datacenter config
+	if err := v.SetupEnvsAndDefaults(ctx, dataCenterConfig); err != nil {
+		v.Log.Error(err, "Failed to set up env vars and default values for VsphereDatacenterConfig")
+		return reconciler.Result{}, err
+	}
+
+	machineConfigMap := map[string]*anywherev1.VSphereMachineConfig{}
+
+	if cluster.Spec.ExternalEtcdConfiguration != nil {
+		mc := &anywherev1.VSphereMachineConfig{}
+		mcName := types.NamespacedName{Namespace: cluster.Namespace, Name: cluster.Spec.ExternalEtcdConfiguration.MachineGroupRef.Name}
+		if err := v.Client.Get(ctx, mcName, mc); err != nil {
+			return reconciler.Result{}, err
+		}
+		v.Log.V(4).Info("Using etcd machine config %v", mc)
+		machineConfigMap[cluster.Spec.ExternalEtcdConfiguration.MachineGroupRef.Name] = mc
+	}
+
+	controlplaneMachine := &anywherev1.VSphereMachineConfig{}
+	controlplaneMachineNameName := types.NamespacedName{Namespace: cluster.Namespace, Name: cluster.Spec.ControlPlaneConfiguration.MachineGroupRef.Name}
+	if err := v.Client.Get(ctx, controlplaneMachineNameName, controlplaneMachine); err != nil {
+		return reconciler.Result{}, err
+	}
+	v.Log.V(4).Info("Using cp machine config %v", controlplaneMachine)
+	machineConfigMap[cluster.Spec.ControlPlaneConfiguration.MachineGroupRef.Name] = controlplaneMachine
+
+	workerNodeMachine := &anywherev1.VSphereMachineConfig{}
+	workerNodeMachineName := types.NamespacedName{Namespace: cluster.Namespace, Name: cluster.Spec.WorkerNodeGroupConfigurations[0].MachineGroupRef.Name}
+	if err := v.Client.Get(ctx, workerNodeMachineName, workerNodeMachine); err != nil {
+		return reconciler.Result{}, err
+	}
+	v.Log.V(4).Info("Using wn machine config %v", workerNodeMachine)
+	machineConfigMap[cluster.Spec.WorkerNodeGroupConfigurations[0].MachineGroupRef.Name] = workerNodeMachine
```

Comment:
could we do this in loop over `cluster.Spec.WorkerNodeGroupConfigurations`?
We are gonna support multiple worker node groups very soon, it would save us some time to come and refactor this

I would vote to not make the "one single worker node group" assumption in any of the new code

Reply:
done